File: src/EmDee.f90
!   This file is part of EmDee.
!
!    EmDee is free software: you can redistrc_intute it and/or modify
!    it under the terms of the GNU General Public License as published by
!    the Free Software Foundation, either version 3 of the License, or
!    (at your option) any later version.
!
!    EmDee is distrc_intuted in the hope that it will be useful,
!    but WITHOUT ANY WARRANTY; without even the implied warranty of
!    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
!    GNU General Public License for more details.
!
!    You should have received a copy of the GNU General Public License
!    along with EmDee. If not, see <http://www.gnu.org/licenses/>.
!
!    Author: Charlles R. A. Abreu (abreu@eq.ufrj.br)
!            Applied Thermodynamics and Molecular Simulation
!            Federal University of Rio de Janeiro, Brazil

module EmDee

use iso_c_binding

implicit none

type, bind(C) :: tEmDee
  integer(c_int) :: builds         ! Number of neighbor-list builds
  real(c_double) :: pairTime       ! Time taken in force calculations
  real(c_double) :: totalTime      ! Total time since initialization
  real(c_double) :: Potential      ! Total potential energy of the system
  real(c_double) :: Kinetic        ! Total kinetic energy of the system
  real(c_double) :: Rotational     ! Rotational kinetic energy of the system
  real(c_double) :: Virial         ! Total internal virial of the system
  type(c_ptr)    :: Data           ! Pointer to EmDee system data
end type tEmDee

interface

  type(tEmDee) function EmDee_system( threads, rc, skin, N, types, masses, seed ) &
    bind(C,name="EmDee_system")
    import :: c_int, c_double, c_ptr, tEmDee
    integer(c_int), value :: threads, N, seed
    real(c_double), value :: rc, skin
    type(c_ptr), value :: types, masses
  end function EmDee_system

  subroutine EmDee_set_charges( md, charges ) &
    bind(C,name="EmDee_set_charges")
    import :: tEmDee, c_ptr
    type(tEmDee), value :: md
    type(c_ptr), value :: charges
  end subroutine EmDee_set_charges

  subroutine EmDee_set_pair_type( md, itype, jtype, model ) &
    bind(C,name="EmDee_set_pair_type")
    import :: tEmDee, c_ptr, c_int
    type(tEmDee), value :: md
    integer(c_int), value :: itype, jtype
    type(c_ptr), value :: model
  end subroutine EmDee_set_pair_type

  subroutine EmDee_ignore_pair( md, i, j ) &
    bind(C,name="EmDee_ignore_pair")
    import :: tEmDee, c_ptr, c_int
    type(tEmDee), value :: md
    integer(c_int), value :: i, j
  end subroutine EmDee_ignore_pair

  subroutine EmDee_add_bond( md, i, j, model ) &
    bind(C,name="EmDee_add_bond")
    import :: tEmDee, c_ptr, c_int
    type(tEmDee), value :: md
    integer(c_int), value :: i, j
    type(c_ptr), value :: model
  end subroutine EmDee_add_bond

  subroutine EmDee_add_angle( md, i, j, k, model ) &
    bind(C,name="EmDee_add_angle")
    import :: tEmDee, c_ptr, c_int
    type(tEmDee), value :: md
    integer(c_int), value :: i, j, k
    type(c_ptr), value :: model
  end subroutine EmDee_add_angle

  subroutine EmDee_add_dihedral( md, i, j, k, l, model ) &
    bind(C,name="EmDee_add_dihedral")
    import :: tEmDee, c_ptr, c_int
    type(tEmDee), value :: md
    integer(c_int), value :: i, j, k, l
    type(c_ptr), value :: model
  end subroutine EmDee_add_dihedral

  subroutine EmDee_add_rigid_body( md, N, indexes ) &
    bind(C,name="EmDee_add_rigid_body")
    import :: tEmDee, c_ptr, c_int
    type(tEmDee), value :: md
    type(c_ptr), value :: indexes
    integer(c_int), value :: N
  end subroutine EmDee_add_rigid_body

  subroutine EmDee_upload( md, Lbox, coords, momenta, forces ) &
    bind(C,name="EmDee_upload")
    import :: tEmDee, c_ptr
    type(tEmDee), intent(inout) :: md
    type(c_ptr), value :: Lbox, coords, momenta, forces
  end subroutine EmDee_upload

  subroutine EmDee_download( md, Lbox, coords, momenta, forces ) &
    bind(C,name="EmDee_download")
    import :: tEmDee, c_ptr
    type(tEmDee), value :: md
    type(c_ptr), value :: Lbox, coords, momenta, forces
  end subroutine EmDee_download

  subroutine EmDee_random_momenta( md, kT, adjust ) &
    bind(C,name="EmDee_random_momenta")
    import :: tEmDee, c_ptr, c_double, c_int
    type(tEmDee), intent(inout) :: md
    real(c_double), value :: kT
    integer(c_int), value :: adjust
  end subroutine EmDee_random_momenta

  subroutine EmDee_boost( md, lambda, alpha, dt, t_flag, r_flag ) &
    bind(C,name="EmDee_boost")
    import :: tEmDee, c_ptr, c_double, c_int
    type(tEmDee), intent(inout) :: md
    real(c_double), value :: lambda, alpha, dt
    integer(c_int), value :: t_flag, r_flag
  end subroutine EmDee_boost

  subroutine EmDee_move( md, lambda, alpha, dt ) &
    bind(C,name="EmDee_move")
    import :: tEmDee, c_ptr, c_double
    type(tEmDee), intent(inout) :: md
    real(c_double), value :: lambda, alpha, dt
  end subroutine EmDee_move

  subroutine EmDee_compute( md ) &
    bind(C,name="EmDee_compute")
    import :: tEmDee
    type(tEmDee), intent(inout) :: md
  end subroutine EmDee_compute

  subroutine EmDee_group_energy( md, na, atoms, ne, energies ) &
    bind(C,name="EmDee_group_energy")
    import :: tEmDee, c_int, c_ptr
    type(tEmDee), value :: md
    integer(c_int), value :: na, ne
    type(c_ptr), value :: atoms, energies
  end subroutine EmDee_group_energy

  function EmDee_pair_none( ) result(model) &
    bind(C,name="EmDee_pair_none")
    import :: c_ptr
    type(c_ptr) :: model
  end function EmDee_pair_none

  function EmDee_pair_lj( epsilon, sigma ) result(model) &
    bind(C,name="EmDee_pair_lj")
    import :: c_double, c_ptr
    real(c_double), value :: epsilon, sigma
    type(c_ptr) :: model
  end function EmDee_pair_lj

  function EmDee_pair_lj_sf( epsilon, sigma, cutoff ) result(model) &
    bind(C,name="EmDee_pair_lj_sf")
    import :: c_double, c_ptr
    real(c_double), value :: epsilon, sigma, cutoff
    type(c_ptr) :: model
  end function EmDee_pair_lj_sf

  function EmDee_bond_harmonic( k, r0 ) result(model) &
    bind(C,name="EmDee_bond_harmonic")
    import :: c_double, c_ptr
    real(c_double), value :: k, r0
    type(c_ptr) :: model
  end function EmDee_bond_harmonic

  function EmDee_bond_morse( D, alpha, r0 ) result(model) &
    bind(C,name="EmDee_bond_morse")
    import :: c_double, c_ptr
    real(c_double), value :: D, alpha, r0
    type(c_ptr) :: model
  end function EmDee_bond_morse

  function EmDee_angle_harmonic( k, theta0 ) result(model) &
    bind(C,name="EmDee_angle_harmonic")
    import :: c_double, c_ptr
    real(c_double), value :: k, theta0
    type(c_ptr) :: model
  end function EmDee_angle_harmonic

  function EmDee_dihedral_harmonic( k, phi0 ) result(model) &
    bind(C,name="EmDee_dihedral_harmonic")
    import :: c_double, c_ptr
    real(c_double), value :: k, phi0
    type(c_ptr) :: model
  end function EmDee_dihedral_harmonic

end interface

end module EmDee

File: src/EmDee.jl
module EmDee

#TODO: REMOVE THIS DEBUG MESSAGE
  print("\n","running test julia from", pwd(), "\n","\n")

#TODO: 1) should this be here or elsewhere? note that the print runs twice in the terminal when running the test.
#TODO: 2) how to choose the path i.e. relative or absolute, hardcoded or read from file or passed from some makefile DEFINE.
  EmDee_libvar=Libdl.dlopen("./lib/libemdee.so")

  macro dlsym(func, lib) #REF: http://docs.julialang.org/en/release-0.5/manual/calling-c-and-fortran-code/
      z, zlocal = gensym(string(func)), gensym()
      eval(current_module(),:(global $z = C_NULL))
      z = esc(z)
      quote
          let $zlocal::Ptr{Void} = $z::Ptr{Void}
              if $zlocal == C_NULL
                 $zlocal = Libdl.dlsym($(esc(lib))::Ptr{Void}, $(esc(func)))
                 global $z = $zlocal
              end
              $zlocal
          end
      end
  end

  type tEmDee
    builds::Int32           # Number of neighbor-list builds
    pairTime::Float64       # Time taken in force calculations
    totalTime::Float64      # Total time since initialization
    Potential::Float64      # Total potential energy of the system
    Kinetic::Float64        # Total kinetic energy of the system
    Rotational::Float64     # Rotational kinetic energy of the system
    Virial::Float64         # Total internal virial of the system
    DOF::Int32              # Total number of degrees of freedom
    RDOF::Int32             # Number of rotational degrees of freedom
    rotationMode::Int32     # Algorithm used for free rotation of rigid bodies
    Data::Ptr{Void}         # Pointer to EmDee system data
  end

  typealias VecOrPtr Union{Vector{Float64},Ptr{Void}}
  typealias MatOrPtr Union{Matrix{Float64},Ptr{Void}}

#---------------------------------------------------------------------------------------------------
function system( threads::Int, layers::Int, rc::Float64, skin::Float64, N::Int,
                 types::Vector{Int}, masses::VecOrPtr )
  return ccall( @dlsym("EmDee_system", EmDee_libvar) tEmDee,
                (Int32, Int32, Float64, Float64, Int32, Ptr{Int32}, Ptr{Float64}),
                threads, layers, rc, skin, N, Vector{Int32}(types), masses )
end

function system( threads::Int, layers::Int, rc::Float64, skin::Float64, N::Int,
                 types::Ptr{Void}, masses::VecOrPtr )
  return ccall( @dlsym("EmDee_system", EmDee_libvar) tEmDee,
                (Int32, Int32, Float64, Float64, Int32, Ptr{Int32}, Ptr{Float64}),
                threads, layers, rc, skin, N, types, masses )
end
#---------------------------------------------------------------------------------------------------
  function set_pair_type( md::tEmDee, itype::Int, jtype::Int, model::Ptr{Void} )
    ccall( @dlsym("EmDee_set_pair_type", EmDee_libvar), Void,
           (tEmDee, Int32, Int32, Ptr{Void}),
           md, itype, jtype, model )
  end
#---------------------------------------------------------------------------------------------------
  function switch_model_layer( md::tEmDee, layer::Int )
    ccall( @dlsym("EmDee_switch_model_layer", EmDee_libvar), Void, (tEmDee, Int32), md, layer )
  end
#---------------------------------------------------------------------------------------------------
  function set_charges( md::tEmDee, charges::Vector{Float64} )
    ccall( @dlsym("EmDee_set_charges", EmDee_libvar), Void, (tEmDee, Ptr{Float64}), md, charges )
  end
#---------------------------------------------------------------------------------------------------
  function add_bond( md::tEmDee, i::Int, j::Int, model::Ptr{Void} )
    ccall( @dlsym("EmDee_add_bond", EmDee_libvar), Void,
           (tEmDee, Int32, Int32, Ptr{Void}), md, i, j, model )
  end
#---------------------------------------------------------------------------------------------------
  function add_angle( md::tEmDee, i::Int, j::Int, k::Int, model::Ptr{Void} )
    ccall( @dlsym("EmDee_add_angle", EmDee_libvar), Void,
           (tEmDee, Int32, Int32, Int32, Ptr{Void}), md, i, j, k, model )
  end
#---------------------------------------------------------------------------------------------------
  function add_dihedral( md::tEmDee, i::Int, j::Int, k::Int, l::Int, model::Ptr{Void} )
    ccall( @dlsym("EmDee_add_dihedral", EmDee_libvar), Void,
           (tEmDee, Int32, Int32, Int32, Int32, Ptr{Void}), md, i, j, k, l, model )
  end
#---------------------------------------------------------------------------------------------------
  function add_rigid_body( md::tEmDee, N::Int, indexes::Vector{Int} )
    ccall( @dlsym("EmDee_add_rigid_body", EmDee_libvar), Void,
           (tEmDee, Int32, Ptr{Int32}), md, N, indexes )
  end
#---------------------------------------------------------------------------------------------------
  function ignore_pair( md::tEmDee, i::Int, j::Int )
    ccall( @dlsym("EmDee_ignore_pair", EmDee_libvar), Void, (tEmDee, Int32, Int32), md, i, j )
  end
#---------------------------------------------------------------------------------------------------
  function upload( md::tEmDee, Lbox::VecOrPtr, coords::MatOrPtr,
                   momenta::MatOrPtr, forces::MatOrPtr )
    ccall( @dlsym("EmDee_upload", EmDee_libvar), Void,
           (Ptr{Void}, Ptr{Float64}, Ptr{Float64}, Ptr{Float64}, Ptr{Float64}),
           pointer_from_objref(md), Lbox, coords, momenta, forces )
  end
#---------------------------------------------------------------------------------------------------
  function download( md::tEmDee, Lbox::VecOrPtr, coords::MatOrPtr,
                     momenta::MatOrPtr, forces::MatOrPtr )
    ccall( @dlsym("EmDee_download", EmDee_libvar), Void,
           (tEmDee, Ptr{Float64}, Ptr{Float64}, Ptr{Float64}, Ptr{Float64}),
           md, Lbox, coords, momenta, forces )
  end
#---------------------------------------------------------------------------------------------------
  function random_momenta( md::tEmDee, kT::Float64, adjust::Int, seed::Int )
    ccall( @dlsym("EmDee_random_momenta", EmDee_libvar), Void,
           (Ptr{Void}, Float64, Int32, Int32),
           pointer_from_objref(md), kT, adjust, seed )
  end
#---------------------------------------------------------------------------------------------------
  function boost( md::tEmDee, lambda::Float64, alpha::Float64, dt::Float64,
                        translation::Int, rotation::Int )
    ccall( @dlsym("EmDee_boost", EmDee_libvar), Void,
           (Ptr{Void}, Float64, Float64, Float64, Int32, Int32),
           pointer_from_objref(md), lambda, alpha, dt, translation, rotation )
  end
#---------------------------------------------------------------------------------------------------
  function move( md::tEmDee, lambda::Float64, alpha::Float64, dt::Float64 )
    ccall( @dlsym("EmDee_move", EmDee_libvar), Void,
           (Ptr{Void}, Float64, Float64, Float64),
           pointer_from_objref(md), lambda, alpha, dt )
  end
#---------------------------------------------------------------------------------------------------
#                                            M O D E L S
#---------------------------------------------------------------------------------------------------
  function pair_lj( ɛ::Float64, σ::Float64 )
    return ccall( @dlsym("EmDee_pair_lj", EmDee_libvar), Ptr{Void}, (Float64, Float64), ɛ, σ )
  end
#---------------------------------------------------------------------------------------------------
  function pair_lj_sf( ɛ::Number, σ::Number, rc::Number )
    return ccall( @dlsym("EmDee_pair_lj_sf", EmDee_libvar), Ptr{Void},
                  (Float64, Float64, Float64), ɛ, σ, rc )
  end
#---------------------------------------------------------------------------------------------------
  function bond_harmonic( k::Number, r₀::Number )
    return ccall( @dlsym("EmDee_bond_harmonic", EmDee_libvar), Ptr{Void}, (Float64, Float64), k, r₀ )
  end
#---------------------------------------------------------------------------------------------------
  function bond_morse( D::Number, α::Number, r₀::Number )
    return ccall( @dlsym("EmDee_bond_morse", EmDee_libvar), Ptr{Void},
                  (Float64, Float64, Float64), D, α, r₀ )
  end
#---------------------------------------------------------------------------------------------------
  function angle_harmonic( k::Number, θ₀::Number )
    return ccall( @dlsym("EmDee_angle_harmonic", EmDee_libvar), Ptr{Void}, (Float64, Float64), k, θ₀ )
  end
#---------------------------------------------------------------------------------------------------
  function dihedral_harmonic( k::Number, ϕ₀::Number )
    return ccall( @dlsym("EmDee_dihedral_harmonic", EmDee_libvar), Ptr{Void}, (Float64, Float64), k, ϕ₀ )
  end
#---------------------------------------------------------------------------------------------------
end
